main/src/main/java/cgeo/geocaching/utils/OfflineTranslateUtils.java

```diff
@@ -127,21 +132,38 @@ public static void getTranslator(final Activity activity, final Language sourceL
     }
 
     public static void translateParagraph(final Translator translator, final OfflineTranslateUtils.Status status, final String text, final Consumer<String> consumer, final Consumer<Exception> errorConsumer) {
-        final List<String> origText = Arrays.asList(text.split("\n"));
-        final List<String> translatedText = new ArrayList<>();
-        for (String p : origText) {
-            translator.translate(p)
-                    .addOnSuccessListener(translation -> {
-                        translatedText.add(translation);
-                        if (origText.size() == translatedText.size()) {
-                            consumer.accept(String.join("\n", translatedText));
-                            status.updateProgress();
-                        }
-                    })
-                    .addOnFailureListener(e -> {
-                        status.abortTranslation();
-                        errorConsumer.accept(e);
-                    });
+        final Document document = Jsoup.parse(text);
+        final Elements elements = document.children().select("*");
+        final AtomicInteger remaining = new AtomicInteger(elements.size());
+        for (Element element : elements) {
+            final String current = element.ownText();
+            if (StringUtils.isBlank(current)) {
```

Comment:
Ignoring blanks is maybe causing missing paragraphs in result. Maybe add a `\n` for each blank? (Just guessing here)

Reply:
I'll dive into it to see where the line breaks get lost, either in the first branch here, or whether `wholeOwnText()` in the second branch will help.